### ISSUE TITLE
[Feature Request] Automatically turn hashtags to tags

### ISSUE THREAD

**Issue by pablo03v** [cryptee/web-client#206]
**Is your feature request related to a problem? Please describe.**
I tried importing markdown files from logseq, but need to convert all tags I created inside of logseq to actual tags in cryptee.

Logseq also uses a special version where they use [example] instead of #example so that names with spaces can be tags. I'd love to see this in cryptee as well but that's not a biggie.

**Comment by johnozbay** [comment]
Hey @pabloscloud! 👋🏻

Thanks for filing this 🙏🏻

— Apologies for the delayed responses here, we have been _extremely_ busy since the whole EU / US vs Apple saga began and I had to testify in March. 

[I wrote a little bit about Markdown](https://github.com/cryptee/web-client/issues/204#issuecomment-2114209339) in #204 — specifically about how `[xyz]` style doc-to-doc links / tags annotation etc is non-standard and [not a part of the official markdown spec](https://spec.commonmark.org/0.31.2/#link-reference-definitions). 

And since Markdown doesn't support tons and tons of features Cryptee has (tags, tables, font sizes you name it), we don't use Markdown as a storage format, because it wouldn't be interoperable or safe to do so. 

We can add a custom tag format by hand and extend on our own shortcuts _while typing in stuff_ though. (**importantly however**, _this wouldn't work while importing/converting markdown files_ to cryptee docs — it would only work while typing in _realtime_. Making sure our importer works would require us to build upon the markdown spec, since our importer is following the markdown spec for interoperability reasons. 

(and occasionally we cheat to allow github flavored markdown (GFM) specs, i.e. in cases of tables)
https://github.com/cryptee/web-client/blob/98e8c3900a3a89b9036571c89ecf4649272a4fb0/source/docs.kit#L225

Let me know what you think, and we can consider adding in another shortcut to tags if that would help your use case — even though these tags won't be visible on export to markdown.

Hoping this makes sense. 

All the best,

J

**Comment by pablo03v** [comment]
I get that you don't use markdown to store notes, but when importing a markdown file (either by uploading or copy pasting) cryptee could identify the hashtags in the text and convert words to tags, right? I don't care too much about these bracket tags (which just added the functionality of having multiple words as one tag).

Hope that makes sense! I appreciate all your work towards apple drama and the public chats with Henry from Techlore on yt ;)

**Comment by johnozbay** [comment]
Thanks for getting back and your kind words! 

It's a little trickier, but we maaaay be able to do this. 
The difficulty is, think of our importer and our realtime-detector as two separate engines. 
The thing that bulk-parses text is not the same as the thing that listens for you type keys and trigger shortcuts 

i.e. when you press `#` it triggers the app to start listening like : ok the keys that are about to get pressed now are the tag, until user presses `spacebar` or `return`. 

But when you bulk import, you don't go through the text character by character, as that would be insanely inefficient. 

And if we do simple regex detection for things like `#` symbols etc, it opens the doors to two problems. 
1) it quickly becomes a lot of work to go through character by character, for each edge cases. 
i.e. 
`#` ->`space` -> `text`  => H1 TEXT, 
`##` -> `space`-> `text` => H2 TEXT etc
So you need to go through all possible scenarios many times for each combination the editor. Any tiny change that deviates from markdown standard = means you'll have to maintain a parallel markdown standard spec for life now. 

2) it makes it possible for people to try and attempt XSS attacks by using `#` symbols. We utilize DOMPurify when importing HTML / Text / Markdown etc into Cryptee, so that you can't import a malicious HTML/TXT/MD file into Cryptee that could attempt to steal your keys. [And the first rule of XSS prevention is : don't sanitize first, then modify / make changes.](https://github.com/cure53/DOMPurify?tab=readme-ov-file#is-there-any-foot-gun-potential) 
(i.e. in this case, change hashtags to something else _after_ sanitization & markdown import) 

So I'm going to look into this and see if we can detect `#` tags _before_ we import markdown / sanitize. But, this means inserting our custom tag-code into the markdown, then recovering it from the markdown parser. (and hope that it doesn't break it) But no promises. 

TLDR; Point being : 
Markdown spec doesn't support tags. 
If you want tags and data portability you should use HTML / ECD / UECD files. 

All implementations of markdown you see in the wild that support tags are custom, random, non-compliant, not-interoperable implementations made by others. Even here on Github, in this text editor. If you use `#` it makes it possible to link issues to one another. [YET, this feature is _not_ a part of github flavored markdown spec](https://github.github.com/gfm/).

Hoping this makes sense! 
Will keep this thread posted once we have more info ✌🏻
